File: src/Shader.cpp
/*
  Copyright (C) 2009 Jon Macey

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

//#include <QFile>
#include "Shader.h"
#include <fstream>
#include <iostream>
#include <memory>
#include <pystring.h>
#include <string> // needed for windows build as get error otherwise
//----------------------------------------------------------------------------------------------------------------------
/// @file Shader.cpp
/// @brief implementation files for Shader class
//----------------------------------------------------------------------------------------------------------------------

namespace ngl
{
//----------------------------------------------------------------------------------------------------------------------
void printInfoLog(const GLuint &_obj)
{
  GLint infologLength = 0;
  GLint charsWritten = 0;
  std::unique_ptr< char[] > infoLog;
  glGetShaderiv(_obj, GL_INFO_LOG_LENGTH, &infologLength);
  if(infologLength > 0)
  {
    infoLog.reset(new char[infologLength]);
    glGetShaderInfoLog(_obj, infologLength, &charsWritten, infoLog.get());

    NGLMessage::addError(infoLog.get(), TimeFormat::NONE);
  }
}

Shader::Shader(std::string_view _name, ShaderType _type, ErrorExit _exitOnError) noexcept :
m_name{_name},m_errorExit(_exitOnError)
{
  switch(_type)
  {
    case ShaderType::VERTEX:
    {
      m_shaderHandle = glCreateShader(GL_VERTEX_SHADER);
      break;
    }
    case ShaderType::FRAGMENT:
    {
      m_shaderHandle = glCreateShader(GL_FRAGMENT_SHADER);
      break;
    }
    case ShaderType::GEOMETRY:
    {
      m_shaderHandle = glCreateShader(GL_GEOMETRY_SHADER);
      break;
    }
    case ShaderType::TESSCONTROL:
    {
      m_shaderHandle = glCreateShader(GL_TESS_CONTROL_SHADER);
      break;
    }
    case ShaderType::TESSEVAL:
    {
      m_shaderHandle = glCreateShader(GL_TESS_EVALUATION_SHADER);
      break;
    }
    case ShaderType::COMPUTE:
#if defined(__APPLE__)
      NGLMessage::addError("Apple doesn't support Computer Shaders ");
#else
    {
      m_shaderHandle = glCreateShader(GL_COMPUTE_SHADER);
    }
#endif
      break;
    case ShaderType::NONE:
    { 
      // nothing to do
    }
  }
  m_compiled = false;
  m_refCount = 0;
}
Shader::~Shader()
{
  // Note this needs to be with cerr as NGLMessage crashes here
  std::cerr << fmt::format("removing shader {0} \n", m_name); 
  glDeleteShader(m_shaderHandle);

}

bool Shader::compile() noexcept
{
  if(m_source.empty())
  {
    NGLMessage::addError("Warning no shader source loaded");
    return false;
  }

  glCompileShader(m_shaderHandle);

  GLint compileStatus = 0;
  glGetShaderiv(m_shaderHandle, GL_COMPILE_STATUS, &compileStatus);
  m_compiled = static_cast< bool >(compileStatus);
  if(m_debugState)
  {
    NGLMessage::addMessage(fmt::format("Compiling Shader {0}", m_name));
    if(compileStatus == GL_FALSE)
    {
      NGLMessage::addError("Shader compile failed or had warnings ");
      printInfoLog(m_shaderHandle);
      if(m_errorExit == ErrorExit::ON)
        exit(EXIT_FAILURE);
    }
  }
  return m_compiled;
}

bool Shader::editShader(std::string_view _toFind, std::string_view _edit)
{
  if(m_source.length() == 0)
  {
    NGLMessage::addError("No shader source to edit");
    return false;
  }
  if(!m_edited)
  {
    m_source = pystring::replace(m_originalSource, _toFind.data(), _edit.data());
    m_edited = true;
  }
  else
  {
    m_source = pystring::replace(m_source, _toFind.data(), _edit.data());
  }
  const char *data = m_source.c_str();
  glShaderSource(m_shaderHandle, 1, &data, nullptr);
  m_compiled = false;
  return true;
}

void Shader::resetEdits()
{
  m_edited = false;
}

void Shader::load(std::string_view _name) noexcept
{
  // see if we already have some source attached
  if(!m_source.empty())
  {
    NGLMessage::addWarning("deleting existing source code\n");
    m_source.clear();
  }
  std::ifstream shaderSource(_name.data());
  if(!shaderSource.is_open())
  {
    NGLMessage::addError(fmt::format("File not found {0}", _name.data()));
    exit(EXIT_FAILURE);
  }
  // now read in the data
  m_source = std::string((std::istreambuf_iterator< char >(shaderSource)), std::istreambuf_iterator< char >());
  shaderSource.close();
  // glShaderSource needs null terminated const char *
  m_source += '\0';
  m_originalSource = m_source;
  const char *data = m_source.c_str();
  glShaderSource(m_shaderHandle, 1, &data, nullptr);
  m_compiled = false;

  if(m_debugState)
  {
    printInfoLog(m_shaderHandle);
  }
}

void Shader::loadFromString(std::string_view _string) noexcept
{
  // see if we already have some source attached
  if(m_source.size() != 0)
  {
    NGLMessage::addWarning("deleting existing source code\n");
    m_source.clear();
  }

  // we need this for the check in the compile bit
  m_source = _string;
  const char *data = m_source.data();

  glShaderSource(m_shaderHandle, 1, &data, nullptr);
  m_compiled = false;
  if(m_debugState == true)
  {
    printInfoLog(m_shaderHandle);
  }
}

} // namespace ngl

File: py-ngl/src/ngl/shader.py
from __future__ import annotations

from enum import Enum

import OpenGL.GL as gl
from .log import logger


class ShaderType(Enum):
    VERTEX = gl.GL_VERTEX_SHADER
    FRAGMENT = gl.GL_FRAGMENT_SHADER
    GEOMETRY = gl.GL_GEOMETRY_SHADER
    TESSCONTROL = gl.GL_TESS_CONTROL_SHADER
    TESSEVAL = gl.GL_TESS_EVALUATION_SHADER
    COMPUTE = gl.GL_COMPUTE_SHADER
    NONE = -1


class MatrixTranspose(Enum):
    TransposeOn = gl.GL_TRUE
    TransposeOff = gl.GL_TRUE


class Shader:
    def __init__(self, name: str, type: int, exit_on_error: bool = True):
        self._name = name
        self._type = type
        self._exit_on_error = exit_on_error
        self._id = gl.glCreateShader(type)
        self._source = ""

    def load(self, source_file: str):
        with open(source_file, "r") as f:
            self._source = f.read()
        gl.glShaderSource(self._id, self._source)

    def compile(self) -> bool:
        gl.glCompileShader(self._id)
        if gl.glGetShaderiv(self._id, gl.GL_COMPILE_STATUS) != gl.GL_TRUE:
            info = gl.glGetShaderInfoLog(self._id)
            logger.error(f"Error compiling shader {self._name=}: {info=}")
            if self._exit_on_error:
                exit()
            return False
        return True

    def edit_shader(self, to_find: str, replace_with: str) -> bool:
        if self._source:
            self._source = self._source.replace(to_find, replace_with)
            gl.glShaderSource(self._id, self._source)
            return True
        return False

    def reset_edits(self):
        if self._source:
            gl.glShaderSource(self._id, self._source)

    def load_shader_source_from_string(self, shader_source: str):
        self._source = shader_source


# class ShaderProgram:
#     def __init__(self, name: str, exit_on_error: bool = True):
#         self._name = name
#         self._exit_on_error = exit_on_error
#         self._id = gl.glCreateProgram()
#         self._shaders = []
#         self._uniforms = {}

#     def attach_shader(self, shader: Shader):
#         gl.glAttachShader(self._id, shader._id)
#         self._shaders.append(shader)

#     def link(self) -> bool:
#         gl.glLinkProgram(self._id)
#         if gl.glGetProgramiv(self._id, gl.GL_LINK_STATUS) != gl.GL_TRUE:
#             info = gl.glGetProgramInfoLog(self._id)
#             print(f"Error linking program {self._name}: {info}")
#             if self._exit_on_error:
#                 exit()
#             return False
#         return True

#     def use(self):
#         gl.glUseProgram(self._id)

#     def get_id(self) -> int:
#         return self._id

#     def get_uniform_location(self, name: str) -> int:
#         if name not in self._uniforms:
#             self._uniforms[name] = gl.glGetUniformLocation(self._id, name)
#         return self._uniforms[name]

#     def set_uniform(self, name: str, *value):
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             if len(value) == 1:
#                 if isinstance(value[0], int):
#                     gl.glUniform1i(loc, value[0])
#                 elif isinstance(value[0], float):
#                     gl.glUniform1f(loc, value[0])
#                 else:
#                     try:
#                         val = list(value[0])
#                         if len(val) == 4:
#                             gl.glUniformMatrix2fv(loc, 1, gl.GL_TRUE, (ctypes.c_float * 4)(*val))
#                         elif len(val) == 9:
#                             gl.glUniformMatrix3fv(loc, 1, gl.GL_TRUE, (ctypes.c_float * 9)(*val))
#                         elif len(val) == 16:
#                             gl.glUniformMatrix4fv(loc, 1, gl.GL_TRUE, (ctypes.c_float * 16)(*val))
#                     except TypeError:
#                         pass
#             elif len(value) == 2:
#                 gl.glUniform2f(loc, *value)
#             elif len(value) == 3:
#                 gl.glUniform3f(loc, *value)
#             elif len(value) == 4:
#                 gl.glUniform4f(loc, *value)

#     def get_uniform_1f(self, name: str) -> float:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 1)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return result[0]
#         return 0.0

#     def get_uniform_2f(self, name: str) -> list[float]:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 2)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return list(result)
#         return [0.0, 0.0]

#     def get_uniform_3f(self, name: str) -> list[float]:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 3)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return list(result)
#         return [0.0, 0.0, 0.0]

#     def get_uniform_4f(self, name: str) -> list[float]:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 4)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return list(result)
#         return [0.0, 0.0, 0.0, 0.0]

#     def get_uniform_mat2(self, name: str) -> list[float]:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 4)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return list(result)
#         return [0.0] * 4

#     def get_uniform_mat3(self, name: str) -> list[float]:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 9)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return list(result)
#         return [0.0] * 9

#     def get_uniform_mat4(self, name: str) -> list[float]:
#         loc = self.get_uniform_location(name)
#         if loc != -1:
#             result = (ctypes.c_float * 16)()
#             gl.glGetUniformfv(self._id, loc, result)
#             return list(result)
#         return [0.0] * 16
